### ISSUE TITLE
[next] refactor <Environment> and split it into two components - a multifile version and a single file version

### ISSUE THREAD

**Issue by joshwashywash** [threlte/threlte#1217]
## background

so i'm trying to refactor `<Environment>` cuz it's still in svelte 4 but it's super overloaded to handle different image formats. some formats take one file, for example hdr and exr and others are cube maps where you pass in 6 images into the `files` prop. There is slight overlap with the loaders for these formats but not enough because if you try to derive the loader and call `.load` the type of the first argument is `string & string[]`.

![Screenshot 2024-11-27 at 7 45 28 PM](https://github.com/user-attachments/assets/c9e03950-eb98-4a25-a6ae-3615ee7f7037)

## proposal

what do you all think about splitting `<Environment>` into `<CubeEnvironment>` and `<Environment>`. (not sure if i like these names so its open for suggestion) a cube map version and a single file version. someone with more knowledge can tell me if it's at all common to switch between environments using a cube map with 6 separate images and one that uses a single image.

this change would make the component much easier to type and guarantee correctness. right now there's a lot of `as whatever` casting going on and a lot of different things need to be checked to narrow down the cases. part of designing a good api is having systems that handle whatever you throw at them but i think there's a tradeoff between a very overloaded component and type correctness, dev ex, readability, and keeping things simple and tractable.

the component props of both `<Environment>` and `<CubeEnvironment` would essentially be the same except `<CubeEnvironment>` 's `file` prop would be `[string, string, string, string, string, string]` instead of `string`. actually `files` on `<Environment>` would be named `file` emphasizing that it only operates with one file. this change would mean that the `load` function for `loaders` in each respective component is typed in such a way that you can guarantee that the type of `files` is what `loader.load` expects - `string` for `<Environment>` or `string[]` for `<CubeEnvironment>`. you can also guarantee that `files` has 6 entries in it in the case of `<CubeEnvironment>`, in other words that `files` is a six-tuple.

the way `<Environment>` is implemented now is similar to if `PerspectiveCamera` and `OrthographicCamera` were squished all into one `Camera` class. maybe not to that degree but it's a good analogy 😏 .

## drei's \<Environment>

regarding drei, they stick everything into `<Environment>` but this is just like the case i made for splitting up `GradientTexture`  into `<RadialGradientTexture>` and `<LinearGradientTexture>`. i think this is another place where we should deviate from drei.

## additional benefits

this alteration should make it easier to extend the environment functionality. one thing i think would be cool is to dynamically apply the Environment to different scenes.

`<Environment scene={sceneToApplyEnvTo} />`

## downsides of such a change

the functionality we lose with splitting them up in such a way is that a user can't as easily swap a "cube environment" for a single texture environment. right now this is possible

```svelte
<script lang="ts">
  type CubeMapTuple = [string, string, string, string, string, string];
  const multiple: CubeMapTuple = [
    'px.png',
    'nx.png',
    'py.png',
    'ny.png',
    'pz.png',
    'nz.png'
  ];
  const single = 'singleTexture.png';

// imagine useCubeFiles can be toggled on and off by some event
const files = $derive(useCubeFiles ? cubeMap : single);
</script>

<Environment {files} />
```

in the proposed change above this would become

```svelte
<script lang="ts">
  type CubeMapTuple = [string, string, string, string, string, string];
  const files: CubeMapTuple = [
    'px.png',
    'nx.png',
    'py.png',
    'ny.png',
    'pz.png',
    'nz.png'
  ];
  const file = 'singleTexture.png';
</script>

{#if useCubeMap}
  <CubeEnvironment {files} />
{:else}
  <Environment {file} />
{/if}
```

## other goals of the refactor

> [!NOTE]
> this section may be updated from time to time as a pr for this change is in the works

- [ ] make something like [this example](https://codesandbox.io/p/sandbox/aquarium-forked-jw26ln?file=/src/App.js:18,39) possible. drei's \<Environment> has a special case when children are used. if children are used it will portal the children into a different scene and applies a cube camera texture to the "portal" scene.
- [x] load textures asynchronously using three's `loader.loadAsync` and threlte's `suspense`.
  - [x] `loadAsync` allows you to pass in an `onProgress` callback. should this be a prop on the component?
- [x] `onloadercreated` function prop to be able to further customize the loader after it is created such as setting its credentials or its crossorigin property. Essentially all the set methods of the [loader](https://threejs.org/docs/#api/en/loaders/Loader). either this or a `loader` prop to be able to use your own loader.
- [x] provide a way for the loaded texture to be modified. either a `ontextureloaded` or a getter for the loaded texture.
- [x] allow user to pass in a texture if they already have the cube texture loaded.

**Comment by raymonddaikon** [comment]
This is as far as I got implementing the `EnvironmentPortal` component. However its' still unable to render the virtual scene. Perhaps I'm misunderstanding fbo render targets, but hopefully this could act as a starting point. ```<script lang="ts">
  import { type Snippet } from 'svelte'
  import { useTask, useThrelte } from '@threlte/core'
  import { CubeCamera, Portal, useFBO } from '@threlte/extras'
  import { Euler, Scene, Texture } from 'three'

  type EnvironmentProps = {
    children?: Snippet
    frames?: number
    near?: number
    far?: number
    resolution?: number
    background?: boolean | 'only'
    backgroundBlurriness?: number
    backgroundIntensity?: number
    backgroundRotation?: Euler
    environmentIntensity?: number
    environmentRotation?: Euler
    map?: Texture
    scene?: Scene
    ground?:
      | boolean
      | {
          radius?: number
          height?: number
          scale?: number
        }
  }

  let {
    children,
    near = 0.1,
    far = 1000,
    resolution = 256,
    frames = 1,
    map,
    background = false,
    backgroundBlurriness = 0,
    backgroundIntensity = 1,
    backgroundRotation = new Euler(0, 0, 0),
    environmentIntensity = 1,
    environmentRotation = new Euler(0, 0, 0),
    scene: sceneProp,
  }: EnvironmentProps = $props()

  const { camera, scene, renderer, size } = useThrelte()

  const renderTarget = useFBO($size.width, $size.height, {
    samples: 4,
  })
  const virtualScene = new Scene()

  let cubeCameraRef: CubeCamera

  $effect.pre(() => {
    virtualScene.backgroundBlurriness = backgroundBlurriness
    virtualScene.backgroundIntensity = backgroundIntensity
    virtualScene.backgroundRotation = backgroundRotation
    virtualScene.environmentIntensity = environmentIntensity
    virtualScene.environmentRotation = environmentRotation

    if (background !== 'only') {
      virtualScene.environment = renderTarget.texture
    }
    if (background) {
      virtualScene.background = renderTarget.texture
    }

    renderer.autoClear = false
    if (frames === 1) {
      const autoClear = renderer.autoClear
      renderer.autoClear = true
      cubeCameraRef?.update(renderer, virtualScene)
      renderer.autoClear = autoClear
    }
    cubeCameraRef?.update(renderer, virtualScene)
  })

  let count = $state(1)
  useTask(() => {
    if (frames === Infinity || count < frames) {
      const autoClear = renderer.autoClear
      renderer.autoClear = true
      cubeCameraRef?.update(renderer, virtualScene)
      renderer.autoClear = autoClear
      count++
    }
  })
</script>

<Portal object={virtualScene}>
  {@render children?.()}
  <CubeCamera bind:ref={cubeCameraRef} args={[near, far, renderTarget]} />
</Portal>
```

**Comment by jkukatzki** [comment]
not a huge fan of splitting the component up into two, simply because more components = more confusion for new adaptors imo, but for what it's worth I don't think switching between equirectangular maps and cube maps is a common occurrence so even if it is split up the component switching in the if else block seems totally viable. glad youre making some progress, especially regarding the caching parts

**Comment by DefinitelyMaybe** [comment]
> doesnt one canvas only hold one scene and is a child of ?

There is a default `new THREE.Scene()` created but there is no set rule. Think more in terms of vanilla three.js. Most applications use only one scene but nothing stopping folks from creating multiple and switching between them.

There's also the [view component](https://github.com/threlte/threlte/pull/1208) which is not too far away. It creates extra scenes within threlte 😃 

>  [new adopters] & [common occurrences]

Definitely want a good new adopters experience and that weighs on the doc-driven-dev the most. With a good doc page definitely doable. Imo extra's is a melting pot 😄 common or not it's good soup 🤣

**Comment by joshwashywash** [comment]
>not a huge fan of splitting the component up into two, simply because more components = more confusion for new adopters

i agree to some extent. more components may mean more to learn but i think in this case good docs can help. also `<CubeEnvironment>` and `<Environment>` will be almost identical in terms of usage. it shouldn't be too difficult to answer the question

"which <Environment> component am i supposed to use?"

answer: if you have a cube map use `CubeEnvironment`, if you have an equirectangular map, use `Environment`.

**Comment by joshwashywash** [comment]
closed with #1229
